aep/general/0155/aep.md

```diff
@@ -0,0 +1,106 @@
+# Idempotency
+
+It is sometimes useful for an API to have a unique, customer-provided
+identifier for particular requests. This can be useful for several purposes,
+such as de-duplicating requests from parallel processes, ensuring the safety of
+retries, or auditing.
+
+The purpose of idempotency keys is to provide idempotency guarantees: allowing
+the same request to be issued more than once without subsequent calls having
+any effect. In the event of a network failure, the client can retry the
+request, and the server can detect duplication and ensure that the request is
+only processed once.
+
+## Guidance
+
+APIs **may** add a `string idempotency_key` parameter to request messages
+(including those of standard methods) in order to uniquely identify particular
+requests.
+
+```proto
+message CreateBookRequest {
+  // The parent resource where this book will be created.
+  // Format: publishers/{publisher}
+  string parent = 1 [
+    (google.api.field_behavior) = REQUIRED,
+    (google.api.resource_reference) = {
+      child_type: "library.example.com/Book"
+    }];
+
+  // The book to create.
+  Book book = 2 [(google.api.field_behavior) = REQUIRED];
+
+  // A unique identifier for this request. Restricted to 36 ASCII characters.
+  // A random UUID is recommended.
+  // This request is only idempotent if a `idempotency_key` is provided.
+  string idempotency_key = 3 [(google.api.field_info).format = UUID4];
+}
+```
+
+- Providing an idempotency key **must** guarantee idempotency.
+  - If a duplicate request is detected, the server **should** return the
+    response for the previously successful request, because the client most
+    likely did not receive the previous response.
+  - APIs **may** choose any reasonable timeframe for honoring idempotency keys.
+- The `idempotency_key` field **must** be provided on the request message to
+  which it applies (and it **must not** be a field on resources themselves).
+- Idempotency keys **should** be optional.
+- Idempotency keys **should** be able to be UUIDs, and **may** allow UUIDs to
```

Comment:
so if I had an idempotency key that wasn't annotated to a UUID - I'm not sure if a client could take advantage of it . As a client would want to generate the UUID on behalf of the user, but it would have no idea what would be accepted or not.

It would be much better guidance if uuid4 was always required.

Reply:
Are you just suggesting changing the **should** to a **must**, like this?

```suggestion
- Idempotency keys **must** be able to be UUIDs, and **may** allow UUIDs to
```

I agree that would be convenient for client libraries!